### PR TITLE
[MRG] Recurrence quantification analysis (RQA)

### LINKED ISSUE #837: RQA and Smith-Waterman
#### Description
(Forking this issue off from #741 , since it's pretty different from the initial idea there.  Tagging @ctralie for input though.)

librosa does not currently implement much in the way of recurrence quantification analysis, beyond standard DTW.  It would be easy to provide some basic functions for recurrence post-processing, such as the L/S/Q methods described by [Serra et al., 2009](http://iopscience.iop.org/article/10.1088/1367-2630/11/9/093017/meta), and [Smith-Waterman](https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm).

The L-method simply accumulates diagonals, and resets the counter whenever there's a gap (`R[i,j] = 0`).

The S-method accumulates diagonals, but allows for gaps of length 1 by considering the max over the immediate diagonal and knight's moves away from `R[i, j]`.

The Q method is closely related to Smith-Waterman, with the following modifications:
- only gaps of length 1 are directly computed
- instead of diagonal/horizontal/vertical, the gaps are diagonal and knight's moves
- different gap penalties are applied depending on whether an entry introduces a new gap or extends an existing one.

While Q and SW are related, I don't see a clean way to implement both with the same underlying function.  I think it would be simpler to have two independent functions for these.
Q and S can be simply implemented though: as noted by Serra et al., when the gap penalties become large, the Q method simplifies to the S method (all gaps instigate a reset).
Likewise, if we make knight's moves a toggle, then it's easy to restrict Q down to L (with large penalties and knight moves disabled).

As mentioned in the previous thread, I've prototyped these already with numba, and they're pretty fast.  The only remaining challenges are:
1. supporting sparse input
2. making a generic API to encapsulate all of the above methods
3. generalize to accumulate link strength rather than counting steps.  With binary input, this recovers the original implementation.

### PR DESCRIPTION
**Pull request by bmcfee**
#### Reference Issue
Implements #837 
[added by @lostanlen: closes #837]

#### What does this implement/fix? Explain your changes.

This PR adds RQA methods as described in the reference issue.

#### Any other comments?

I think this is ready for CR.

### CODE REVIEW

**Comment by bmcfee** [comment]
For completeness, here's how the docstring examples render:

(No gaps or knight moves)
image

(Gaps and knight moves allowed)
image

File: librosa/sequence.py [hunk 1]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
     return wp
 
 
+def rqa(sim, gap_onset=1, gap_extend=1, knight_moves=True, backtrack=True):
+    '''Recurrence quantification analysis (RQA)
+
+    This function implements different forms of RQA as described by
+    Serra, Serra, and Andrzejak [1]_.  These methods take as input
+    a recurrence or cross-similarity matrix `sim`, and calculate the value
```
Comment (lostanlen) [inline librosa/sequence.py]: ~what is the difference between a recurrence matrix and a cross-similarity matrix?~
also note that, because "recurrence" is a noun, it might not be obvious to non-native English speakers that it is adjectived and that the noun "matrix" applied to both "recurrence" or "cross-similarity".

File: librosa/sequence.py [hunk 2]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+
+    The simplest RQA method, denoted as `L` [1]_ (equation 3) and equivalent
+    to the method described by Eckman, Kamphorst, and Ruelle [2]_, accumulates
+    the length of diagonal paths with positive values in the input:
+
+        - `score[i, j] = score[i-1, j-1] + 1`  if `sim[i, j] > 0`
+        - `score[i, j] = 0` otherwise.
+
+    The second method, denoted as `S` [1]_ (equation 4), is similar to the first,
+    but allows for "knight moves" in addition to strict diagonal moves:
```
Comment (lostanlen) [inline librosa/sequence.py]: would be good to be explicit about the reference to the chess game here.

File: librosa/sequence.py [hunk 3]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+
+        - `score[i, j] = max(score[i-1, j-1], score[i-2, j-1], score[i-1, j-2]) + 1`  if `sim[i, j] > 0`
+        - `score[i, j] = 0` otherwise.
+
+    The third method, denoted as `Q` [1]_ (equations 5 and 6) extends this by
+    allowing gaps in the alignment that incur some cost, rather than a hard
+    reset to 0 whenever `sim[i, j] == 0`.
+    Gaps are penalized by two additional parameters, `gap_onset` and `gap_extend`,
+    which are deducted from the value of the alignment path every time a gap
```
Comment (lostanlen) [inline librosa/sequence.py]: deducted -> subtracted

(i know they are synonyms, but deducted is often confused for deduced by non-native En speakers)

File: librosa/sequence.py [hunk 4]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+    is introduced or extended (respectively).
+
+    Note that setting `gap_onset` and `gap_extend` to `np.inf` recovers the second
+    method, and disabling knight moves recovers the first.
+
+
+    .. [1] Serra, Joan, Xavier Serra, and Ralph G. Andrzejak.
+        "Cross recurrence quantification for cover song identification."
```
Comment (lostanlen) [inline librosa/sequence.py]: Cross-recurrence

File: librosa/sequence.py [hunk 5]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+    reset to 0 whenever `sim[i, j] == 0`.
+    Gaps are penalized by two additional parameters, `gap_onset` and `gap_extend`,
+    which are deducted from the value of the alignment path every time a gap
+    is introduced or extended (respectively).
+
+    Note that setting `gap_onset` and `gap_extend` to `np.inf` recovers the second
+    method, and disabling knight moves recovers the first.
+
+
+    .. [1] Serra, Joan, Xavier Serra, and Ralph G. Andrzejak.
```
Comment (lostanlen) [inline librosa/sequence.py]: technically it's Joan Serrà and Xavier Serra, but i can see why you'd want to asciify Joan's last name into Serra

File: librosa/sequence.py [hunk 6]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+        "Cross recurrence quantification for cover song identification."
+        New Journal of Physics 11, no. 9 (2009): 093017.
```
Comment (lostanlen) [inline librosa/sequence.py]: i don't know what this "093017" means. Apparently it corresponds to the `pages` field in BibTeX ... ?

File: librosa/sequence.py (hunk 1 above)
Comment (lostanlen) [reply librosa/sequence.py]: bis: i would be in favor of replacing recurrence by self-similarity

File: librosa/sequence.py [hunk 7]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+    .. [2] Eckmann, J. P., S. Oliffson Kamphorst, and D. Ruelle.
+        "Recurrence plots of dynamical systems."
+        World Scientific Series on Nonlinear Science Series A 16 (1995): 441-446.
+
+
+    Parameters
+    ----------
+    sim : np.ndarray [shape=(N, M), non-negative]
+        The similarity matrix to use as input.
+
```
Comment (lostanlen) [inline librosa/sequence.py]: unnecessary whitespace

File: librosa/sequence.py [hunk 8]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+    gap_onset : float > 0
+        Penalty for introducing a gap to an alignment sequence
+
+    gap_extend : float > 0
+        Penalty for extending a gap in an alignment sequence
+
+    knight_moves : bool
+        If `True` (default), allow for "knight moves" in the alignment,
+        e.g., `(n, m) => (n + 1, m + 2)` or `(n + 2, m + 1)`.
+
```
Comment (lostanlen) [inline librosa/sequence.py]: unnecessary whitespace

File: librosa/sequence.py [hunk 9]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+        If `False`, only allow for diagonal moves `(n, m) => (n + 1, m + 1)`.
+
+    backtrack : bool
+        If `True`, return the alignment path.
+
```
Comment (lostanlen) [inline librosa/sequence.py]: unnecessary whitespace

File: librosa/sequence.py [hunk 10]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+    Returns
+    -------
+    score : np.ndarray [shape=(N, M)]
+        The alignment score matrix.  `score[n, m]` is the cumulative value of
+        the best alignment sequence ending in frames `n` and `m`.
+
+    path : np.ndarray [shape=(k, 2)] (optional)
+        If `backtrack=True`, `path` contains a list of pairs of aligned frames
+        in the best alignment sequence.
+
```
Comment (lostanlen) [inline librosa/sequence.py]: unnecessary whitespace

File: librosa/sequence.py [hunk 11]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+    # The output array
+    score = np.zeros(sim.shape, dtype=sim.dtype)
+
+    # The backtracking array
+    backtrack = np.zeros(sim.shape, dtype=np.int8)
+
+    # These are place-holder arrays to limit the points being considered
+    # at each step of the DP
+    #
```
Comment (lostanlen) [inline librosa/sequence.py]: unnecessary whitespace (and again later in this comment block)

File: librosa/sequence.py [hunk 12]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+    #  -2 ==> reset with inclusion (ie positive value at init)
+
+    # Initialize the first row and column with the data
+    score[0, :] = sim[0, :]
+    score[:, 0] = sim[:, 0]
+
+    # backtracking initialization: the first row and column are all resets
+    # if there's a positive link here, it's an inclusive reset
+    for i in range(sim.shape[0]):
+        if sim[i, 0]:
```
Comment (lostanlen) [inline librosa/sequence.py]: ⚠️ ⚠️ ⚠️ 
why is it `if sim[i,0]` and not `if sim[i,0]>0`? can `sim` contain `None` or `False`?

File: librosa/sequence.py [hunk 13]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+        # Other bt_index values are okay for inclusion
+        path.insert(0, idx)
+
+        # -2 indicates beginning of sequence,
+        # so we can't backtrack any further
+        if bt_index == -2:
+            break
+
+        # Otherwise, prepend this index and continue
+        idx = [idx[_] + offsets[bt_index][_] for _ in range(len(idx))]
```
Comment (lostanlen) [inline librosa/sequence.py]: is there a reason why you're iterating with the underscore symbol rather than with a letter?

File: librosa/sequence.py [hunk 14]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+
+    if not path:
```
Comment (lostanlen) [inline librosa/sequence.py]: can you put a comment describing when this case typically happens?

File: librosa/sequence.py [hunk 15]
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
...
+    >>> L_score, L_path = librosa.sequence.rqa(rec, np.inf, np.inf,
+    ...                                        knight_moves=False)
+    >>> plt.subplot(1,2,1)
+    >>> librosa.display.specshow(rec, x_axis='frames', y_axis='frames')
+    >>> plt.title('Recurrence matrix')
+    >>> plt.subplot(1,2,2)
+    >>> librosa.display.specshow(L_score, x_axis='frames', y_axis='frames')
+    >>> plt.plot(L_path[:, 1], L_path[:, 0], label='Optimal path', color='c')
+    >>> plt.legend()
+    >>> plt.show()
```
Comment (lostanlen) [inline librosa/sequence.py]: can you make sure that the x-axis and y-axis have equal length? ("square" mode for specshow)

File: librosa/sequence.py (hunk 1 above)
Comment (bmcfee) [reply librosa/sequence.py]: I went with recurrence to match the function name, but we have a see-also, so i'm not too picky about it.

File: librosa/sequence.py (hunk 3 above)
Comment (bmcfee) [reply librosa/sequence.py]: good catch

File: librosa/sequence.py (hunk 5 above)
Comment (bmcfee) [reply librosa/sequence.py]: We do set utf-8 encoding on all source files, so there's not really a problem here.  It may cause some difficulty in docstring rendering in REPL or sphinx, but I'll look into it.

File: librosa/sequence.py (hunk 4 above)
Comment (bmcfee) [reply librosa/sequence.py]: The paper title does not have a hyphen.

File: librosa/sequence.py (hunk 6 above)
Comment (bmcfee) [reply librosa/sequence.py]: I guess so -- this is from the citation export, so I'm inclined to keep it.

File: librosa/sequence.py (hunk 7 above)
Comment (bmcfee) [reply librosa/sequence.py]: where?

File: librosa/sequence.py (hunk 8 above)
Comment (bmcfee) [reply librosa/sequence.py]: where?

File: librosa/sequence.py (hunk 15 above)
Comment (bmcfee) [reply librosa/sequence.py]: specshow doesn't directly support aspect ratios, but i can change the figure size.

File: librosa/sequence.py (hunk 12 above)
Comment (bmcfee) [reply librosa/sequence.py]: sim can't contain none.  It could contain `False` (dtyped to np.bool); in general, nonzero is truthy and zero is not truthy.

File: librosa/sequence.py (hunk 13 above)
Comment (bmcfee) [reply librosa/sequence.py]: code checker does not like single-letter variables; `_` for this purpose is pythonic though.

**Comment by bmcfee** [comment]
Thanks for the CR @lostanlen !  I've addressed most of the issues, and strengthened the tests a little.  Let me know if there's anything else that you think needs changing.

File: librosa/sequence.py (hunk 4 above)
Comment (lostanlen) [reply librosa/sequence.py]: 🤷‍♂

File: librosa/sequence.py (hunk 12 above)
Comment (lostanlen) [reply librosa/sequence.py]: ok i understand now!

**Comment by lostanlen** [comment]
LGTM now! sorry for the false alarms in CR
